src/memray/commands/attach.py

```diff
@@ -0,0 +1,377 @@
+
+tracker = {tracker_call}
+try:
+    tracker.__enter__()
+except:
+    # Prevent the exception from keeping the tracker alive.
+    # This way resources are cleaned up ASAP.
+    del tracker
+    raise
+
+memray._last_tracker = tracker
+"""
+
+
+def inject(debugger: str, pid: int, port: int) -> bool:
+    """Executes a file in a running Python process."""
+    injecter = pathlib.Path(memray.__file__).parent / "_inject.abi3.so"
+    assert injecter.exists()
+
+    gdb_cmd = [
+        "gdb",
+        "-batch",
+        "-p",
+        str(pid),
+        "-nx",
+        "-nw",
+        "-iex=set auto-solib-add off",
+        f"-ex=set $rtld_now={RTLD_NOW}",
+        f'-ex=set $libpath="{injecter}"',
+        f"-ex=set $port={port}",
+        f"-x={GDB_SCRIPT}",
+    ]
+
+    # When adding new breakpoints, also update _attach.gdb
+    lldb_breakpoint_args = [
+        "breakpoint",
+        "set",
+        "-b",
+        "malloc",
+        "-b",
+        "calloc",
+        "-b",
+        "realloc",
+        "-b",
+        "free",
+        "-b",
+        "PyMem_Malloc",
+        "-b",
+        "PyMem_Calloc",
+        "-b",
+        "PyMem_Realloc",
+        "-b",
+        "PyMem_Free",
+        "-C",
+        "'breakpoint disable'",
+        "-C",
+        f"'expr auto $rtld_default = (void*){RTLD_DEFAULT}'",
+        "-C",
+        f"'expr auto $rtld_now = {RTLD_NOW}'",
+        "-C",
+        "'expr auto $dlsym = (void* (*)(void*, const char*))&dlsym'",
+        "-C",
+        """'expr auto $dlopen = $dlsym($rtld_default, "dlopen")'""",
+        "-C",
+        """'expr auto $dlerror = $dlsym($rtld_default, "dlerror")'""",
+        "-C",
+        "'expr auto $dll = ((void*(*)(const char*, int))$dlopen)($libpath, $rtld_now)'",
+        "-C",
+        "'p ((char*(*)(void))$dlerror)()'",
+        "-C",
+        """'expr auto $spawn = $dlsym($dll, "memray_spawn_client")'""",
+        "-C",
+        "'p ((const char*(*)(int))$spawn)($port)'",
+    ]
+
+    lldb_cmd = [
+        "lldb",
+        "--batch",
+        "-p",
+        str(pid),
+        "--no-lldbinit",
+        "--source-quietly",
+        "-o",
+        "p ((void*(*)(size_t))PyMem_Malloc)",
+        "-o",
+        "p ((void*(*)(size_t, size_t))PyMem_Calloc)",
+        "-o",
+        "p ((void*(*)(void *, size_t))PyMem_Realloc)",
+        "-o",
+        "p ((void(*)(void*))PyMem_Free)",
+        "-o",
+        f'expr char $libpath[]="{injecter}"',
+        "-o",
+        f"expr int $port={port}",
+        "-o",
+        " ".join(lldb_breakpoint_args),
+        "-o",
+        "continue",
+    ]
+
+    cmd = gdb_cmd if debugger == "gdb" else lldb_cmd
+    # print(shlex.join(cmd))
+    p = subprocess.Popen(
+        cmd,
+        text=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+    )
+    output = p.communicate()[0]
+    # print(f"debugger return code: {p.returncode}")
+    # print(f"debugger output:\n{output}")
+    return (
+        p.returncode == 0
```

Comment:
I went ahead and did it after all, in a new commit on this PR. I'm not totally happy with it, since it's a bit hacky and relies on parsing error messages from gdb or lldb - but I'm pretty sure anything we do will rely at least somewhat on parsing error messages.